pkg/coordinator/redemptions.go

```diff
@@ -0,0 +1,328 @@
+	}
+
+	if len(redemptionsToPropose) == 0 {
+		return [20]byte{}, nil, nil
+	}
+
+	logger.Infof(
+		"found [%d] redemptions for wallet [%s]",
+		len(redemptionsToPropose),
+		hexutils.Encode(walletPublicKeyHash[:]),
+	)
+
+	redeemersOutputScripts := make([]bitcoin.Script, len(redemptionsToPropose))
+
+	for i, redemption := range redemptionsToPropose {
+		logger.Infof(
+			"redemption [%d/%d] - %s",
+			i+1,
+			len(redemptionsToPropose),
+			fmt.Sprintf(
+				"redemptionKey: [%s], requested at: [%s]",
+				redemption.redemptionKey,
+				redemption.requestedAt,
+			))
+
+		redeemersOutputScripts[i] = redemption.redeemerOutputScript
+	}
+
+	return walletPublicKeyHash, redeemersOutputScripts, nil
+}
+
+// ProposeRedemption handles redemption proposal submission.
+func ProposeRedemption(
+	chain Chain,
+	walletPublicKeyHash [20]byte,
+	fee int64,
+	redeemersOutputScripts []bitcoin.Script,
+	dryRun bool,
+) error {
+	if len(redeemersOutputScripts) == 0 {
+		return fmt.Errorf("redemptions list is empty")
+	}
+
+	// Estimate fee if it's missing.
+	if fee <= 0 {
+		panic("fee estimation not implemented yet")
+	}
+
+	logger.Infof("redemption transaction fee: [%d]", fee)
+
+	logger.Infof("preparing a redemption proposal...")
+
+	proposal := &tbtc.RedemptionProposal{
+		WalletPublicKeyHash:    walletPublicKeyHash,
+		RedeemersOutputScripts: redeemersOutputScripts,
+		RedemptionTxFee:        big.NewInt(fee),
+	}
+
+	logger.Infof("validating the redemption proposal...")
+
+	if _, err := tbtc.ValidateRedemptionProposal(
+		logger,
+		proposal,
+		chain,
+	); err != nil {
+		return fmt.Errorf("failed to verify redemption proposal: %v", err)
+	}
+
+	if !dryRun {
+		logger.Infof("submitting the deposit sweep proposal...")
+		if err := chain.SubmitRedemptionProposalWithReimbursement(proposal); err != nil {
+			return fmt.Errorf("failed to submit deposit sweep proposal: %v", err)
+		}
+	} else {
+		logger.Infof("skipping transaction submission in dry-run mode")
+	}
+
+	return nil
+}
+
+func getPendingRedemptions(
+	chain Chain,
+	walletPublicKeyHash [20]byte,
+	maxNumberOfRedemptions int,
+	redemptionRequestTimeout uint32,
+	redemptionRequestMinAge uint32,
+) ([]*redemptionEntry, error) {
+	logger.Infof("reading revealed deposits from chain...")
+
+	filter := &tbtc.RedemptionRequestedEventFilter{}
+	if walletPublicKeyHash != [20]byte{} {
+		filter.WalletPublicKeyHash = [][20]byte{walletPublicKeyHash}
+	}
+
+	// Only redemption requests in range:
+	// [now - redemptionRequestTimeout, now - redemptionRequestMinAge]
+	// should be taken into consideration.
+	redemptionRequestsRangeStartTimestamp := time.Now().Add(
+		time.Duration(-redemptionRequestTimeout) * time.Second,
+	)
+	redemptionRequestsRangeEndTimestamp := time.Now().Add(
+		time.Duration(-redemptionRequestMinAge) * time.Second,
+	)
+
+	// TODO: We should consider narrowing the block range in filter the fetch
+	//       only events that are newer than `now - redemptionRequestTimeout`.
```

Comment:
This feels like a must-have. Otherwise, the maintainer may stop working after some time.

Reply:
Why may it stop working? We could optimize requests to the Ethereum API by ignoring the block range for the events that were already handled.